### PR TITLE
[Experiment] set .cargo/registry/src as readonly

### PR DESCRIPTION
**Pull request by weihanglo**
Need a crater run on `check` mode to see how readonly registry sources break stuff.

This does not take into account the case which [I mentioned earlier](https://github.com/rust-lang/cargo/issues/9455#issuecomment-1143026498), so failure cases may be far less than expected. I will appreciate if anyone gives a hint about how to make crater run `cargo c && rm -rf ~/.cargo/registry/src` && `cargo c` :)

Still, changing the permission bits causes `build.rs` not able to run some executables[^1], or makes `.cargo-ok` not able to write[^2]. So, let's see how far it goes.

r? @ghost

[^1]: https://github.com/rust-lang/cargo/pull/9131
[^2]: https://github.com/rust-lang/cargo/blob/c9d8c28cba959c347271eaf31c9abfbb74c690bd/src/cargo/sources/registry/mod.rs#L657-L664

### CODE REVIEW

**Comment by weihanglo** [comment]
Thanks for the info! I am fine with check-only mode first. Let me check if I get the permission.

@craterbot run mode=check-only crates=top-100

**Comment by weihanglo** [comment]
Okay, no 😞 
@jyn514 could you help run crater? Thanks!

**Comment by jyn514** [comment]
@craterbot run mode=check-only crates=top-100

**Comment by jyn514** [comment]
@craterbot cancel

@craterbot run mode=check-only crates=top-100 p=1 (this is fast to run, no need to wait a week)

**Comment by jyn514** [comment]
@craterbot abort name=pr-98370

@craterbot run mode=check-only crates=top-100 p=1 (this is fast to run, no need to wait a week)

**Comment by jyn514** [comment]
@craterbot run mode=check-only crates=top-100 p=1 (this is fast to run, no need to wait a week)

**Comment by jyn514** [comment]
@craterbot run mode=check-only crates=top-100 p=1

**Comment by ehuss** [comment]
> I will appreciate if anyone gives a hint about how to make crater run `cargo c && rm -rf ~/.cargo/registry/src` && `cargo c` :)

I haven't had a chance to look at this, but I just wanted to quickly mention that you can implement whatever logic you want inside your cargo fork.  Like in [`commands/check.rs`](https://github.com/rust-lang/cargo/blob/ed549d87178e47cb9a05ba309bd821c9595b770c/src/bin/cargo/commands/check.rs#L43-L56) you can call `fs::remove_dir_all(…)`.  

https://github.com/ehuss/cargo/compare/e96bdb0c3d0a418e7fcd7fbd69be08abf830b4bc...6901690a6f8d519efb4fabf48c1c2b94af0c3bd8 is an example of how I did something similar.  It has an environment variable so that the "outer" `cargo check` would run the logic, and the "inner" cargo (with the environment variable set) would run like normal.

Also, I also wanted to mention that I think crater prefetches all dependencies, so this may not work as expected.  Sorry, I forgot about that when we were discussing this earlier.  I'm also not certain how CARGO_HOME gets mounted in the docker container, and whether or not you can actually modify CARGO_HOME.  (Sorry, I have forgotten a lot of how rustwide works.)

Also, just looking at the diff, it looks like the `src/tools/cargo` submodule may not be at the correct hash. It looks like this has 85e457e158db216a2938d51bc3b617a5a7fe6015, but your branch is at 850d7ce755ad62d9db436ec2697b879d568ce16d. At least, I don't see your changes in this diff.

**Comment by weihanglo** [comment]
Oops. My fault. Thanks for the tip ehuss. Updated here https://github.com/weihanglo/cargo/commit/fa889b338df30a1121207c2fc22fd6de7167ddad

@bors try

**Comment by jyn514** [comment]
@craterbot run mode=check-only crates=top-100 p=1

**Comment by jyn514** [comment]
Looks like @ehuss was right:
```

[INFO] [stderr] >> 1st remove `/opt/rustwide/cargo-home/registry/src`
[INFO] [stderr] error: Read-only file system (os error 30)
```
On the bright side I think this means that anything Crater is compiling isn't writing to the registry?

**Comment by Mark-Simulacrum** [comment]
> On the bright side I think this means that anything Crater is compiling isn't writing to the registry?

I would rephrase this to any *successful* compilation in Crater won't be writing to the registry. You can probably download a full-run's results e.g., https://crater-reports.s3.amazonaws.com/pr-99217/logs-archives/all.tar.gz and grep those logs for Read-only filesystem to try to get an idea of the percentage of crates we fail to run due to this issue (at least at cargo check time).

**Comment by weihanglo** [comment]
Going to close this now. Will revisit if I have more time on it.
